fx-src/FirefoxStorage.js

```diff
@@ -0,0 +1,484 @@
+import BaseAdapter from "../src/adapters/base";
+
+const { interfaces: Ci, results: Cr, utils: Cu } = Components;
+
+Cu.import("resource://gre/modules/Services.jsm");
+Cu.import("resource://gre/modules/FileUtils.jsm");
+
+function debug(message) {
+  // TODO: Read a pref or something to work out whether to emit debug
+  // messages
+  dump("FirefoxStorage "+message+"\n\n");
+}
+
+// A way to handle our storage connection and storage statements
+class KintoStorage {
+  constructor(dbname) {
+    this.dbname = dbname;
+    this.file = FileUtils.getFile("ProfD", ["kinto.sqlite"]);
+    this.statements = {};
+  }
+
+  getStatement(statementString) {
+    debug("getStatement for "+statementString);
+    let statement = this.statements[statementString];
+    if (statement) {
+      debug("statement exists");
+      if (statement.reset) {
+        statement.reset();
+      }
+    } else {
+      debug("statement does not exist; creating");
+      const dbconn = this.getConnection();
+      if (dbconn) {
+        debug("connection is ready? "+dbconn.connectionReady);
+        statement = dbconn.createAsyncStatement(statementString);
+        this.statements[statementString] = statement;
+      } else {
+        debug("unable to get a database connection");
+      }
+    }
+    return statement;
+  }
+
+  getName() {
+    return this.dbname;
+  }
+
+  getConnection() {
+    if (!this._dbconn) {
+      try {
+        this._dbconn = Services.storage.openDatabase(this.file);
```

Comment:
Unless you have a really, really good reason, you should be using Sqlite.jsm instead of Storage. In this review I won't look for Storage footguns.

https://developer.mozilla.org/en-US/docs/Mozilla/JavaScript_code_modules/Sqlite.jsm
http://gregoryszorc.com/blog/2013/04/14/sqlite.jsm---sqlite-done-betterer/